drivers/gpio/gpio_numaker.c

```diff
@@ -0,0 +1,291 @@
+/*
+ * Copyright (c) 2023 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_numaker_gpio
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <soc.h>
+
+#ifdef CONFIG_CLOCK_CONTROL_NUMAKER_SCC
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/clock_control_numaker.h>
+#endif
+#include <zephyr/drivers/gpio/gpio_utils.h>
+#include "NuMicro.h"
+
+#include <zephyr/logging/log.h>
+
+#define NU_MFP_POS(pinindex) ((pinindex % 4) * 8)
+
+LOG_MODULE_REGISTER(gpio_numaker, LOG_LEVEL_ERR);
+
+struct gpio_numaker_config {
+	struct gpio_driver_config common;
+	uint32_t reg;
+	uint32_t gpa_base;
+	uint32_t size;
+	uint32_t clk_modidx;
+#ifdef CONFIG_CLOCK_CONTROL_NUMAKER_SCC
+	const struct device *clk_dev;
+#endif
+};
+
+struct gpio_numaker_data {
+	struct gpio_driver_data common;
+	sys_slist_t callbacks;
+};
+
+static int gpio_numaker_configure(const struct device *dev, gpio_pin_t pin, gpio_flags_t flags)
+{
+	const struct gpio_numaker_config *config = dev->config;
+	struct gpio_numaker_data *data = dev->data;
+	GPIO_T *gpio_base = (GPIO_T *)config->reg;
+	uint32_t pinMfpMask = (0x1f << NU_MFP_POS(pin));
+	uint32_t pinMask = BIT(pin); /* mask for pin index --> (0x01 << pin) */
+	uint32_t port_index;
+	uint32_t *GPx_MFPx;
+	uint32_t pinMfpGpio;
+	int err = 0;
+
+	ARG_UNUSED(data);
+
+	/* Check for an invalid pin number */
+	if (pin >= 15) {
+		return -EINVAL;
+	}
+
+	SYS_UnlockReg();
```

Comment:
can you please detail why we need all these SYS_* calls?

Reply:
In our specific MCU H/W design, we need to unlock H/W before some manipulations. Our MPU chip no this kind control.